### ISSUE TITLE
`SocketException` thrown when receiving heartbeats 

### ISSUE THREAD

**Issue by mfloryan** [rabbitmq/rabbitmq-dotnet-client#90]
When a connection is established to a rabbit server with *heartbeats* enabled the rabbit connection thread regularly throws `SocketException`. It appears the exception is roughly thrown with every other heartbeat message received from the server.
The exception is handled by the driver silently (and thus doesn't affect the client directly) but may indicate an existing bug with the socket use.

The sequence of first chance exceptions is:
```
A first chance exception of type 'System.Net.Sockets.SocketException' occurred in System.dll
A first chance exception of type 'System.IO.IOException' occurred in System.dll
A first chance exception of type 'System.Net.Sockets.SocketException' occurred in RabbitMQ.Client.dll
```

The exception message is: _A connection attempt failed because the connected party did not properly respond after a period of time or established connection failed because connected host has failed to respond._ and it appears that it's the *timeout* part of that reason that causes the exception.

He is a relevant network traffic trace from .Net:
```
System.Net.Sockets Verbose: 0 : [3300] Socket#20234383::Receive()
System.Net.Sockets Verbose: 0 : [3300] Data from Socket#20234383::Receive
System.Net.Sockets Verbose: 0 : [3300] 00000000 : 01 00 01 00 00 00 08 00-14 00 0B 00 00 00 00 CE : ................
System.Net.Sockets Verbose: 0 : [3300] Exiting Socket#20234383::Receive()   -> Int32#16
System.Net.Sockets Verbose: 0 : [3300] Socket#20234383::Receive()
System.Net.Sockets Error: 0 : [3300] Socket#20234383::UpdateStatusAfterSocketError() - TimedOut
System.Net.Sockets Error: 0 : [3300] Exception in Socket#20234383::Receive - A connection attempt failed because the connected party did not properly respond after a period of time or established connection failed because connected host has failed to respond.
System.Net.Sockets Verbose: 0 : [3300] Exiting Socket#20234383::Receive()   -> Int32#0
System.Net.Sockets Verbose: 0 : [3300] Socket#20234383::Receive()
System.Net.Sockets Verbose: 0 : [3300] Data from Socket#20234383::Receive
System.Net.Sockets Verbose: 0 : [3300] 00000000 : 08 00 00 00 00 00 00 CE-                        : ........
System.Net.Sockets Verbose: 0 : [3300] Exiting Socket#20234383::Receive()   -> Int32#8
System.Net.Sockets Verbose: 0 : [3300] Socket#20234383::Receive()
System.Net.Sockets Error: 0 : [3300] Socket#20234383::UpdateStatusAfterSocketError() - TimedOut
System.Net.Sockets Error: 0 : [3300] Exception in Socket#20234383::Receive - A connection attempt failed because the connected party did not properly respond after a period of time or established connection failed because connected host has failed to respond.
System.Net.Sockets Verbose: 0 : [3300] Exiting Socket#20234383::Receive()   -> Int32#0
```

It is however somewhat surprising as a trace similar to the above (with timestamps included) show:

* `2015-05-20T15:05:55.2691853Z` - received OK
* `2015-05-20T15:06:00.3083620Z` - exception in receive
* `2015-05-20T15:06:05.1654930Z` - received OK

all exactly on the same socket.

It appears that every other receive is throwing. Tested on .Net 4.5.2 with .Net driver at version `3.5.2` and server running on linux `3.5.1`

**Comment by michaelklishin** [comment]
Can you post a small example that reproduces the problem as well as RabbitMQ log files?

**Comment by mfloryan** [comment]
This is sufficient to reproduce the problem:

```csharp
using System;
using RabbitMQ.Client;

namespace bunny
{
    class Program
    {
        static void Main(string[] args)
        {
            var factory = new ConnectionFactory
            {
                HostName = "172.16.169.1",
                VirtualHost = "bunny",
                UserName = "bunny",
                Password = "rabbit",
                RequestedHeartbeat = 10
            };
            Console.WriteLine("{0:O}: Starting.", DateTime.UtcNow);
            var connection = factory.CreateConnection();
            Console.WriteLine("{0:O}: Started.", DateTime.UtcNow);
            Console.ReadKey();

            connection.Close();
            Console.WriteLine("{0:O}: Closed.", DateTime.UtcNow);
        }
    }
}
```

And full network trace is created by adding this section to `app.config`

```xml
<system.diagnostics>
  <sources>
    <source name="System.Net" tracemode="includehex" maxdatasize="1024">
      <listeners>
        <add name="System.Net"/>
      </listeners>
    </source>
    <source name="System.Net.Sockets">
      <listeners>
        <add name="System.Net"/>
      </listeners>
    </source>
  </sources>
  <switches>
    <add name="System.Net" value="Verbose"/>
    <add name="System.Net.Sockets" value="Verbose"/>
  </switches>
  <sharedListeners>
    <add name="System.Net"
      type="System.Diagnostics.TextWriterTraceListener"
      traceOutputOptions="ThreadId, DateTime"
      initializeData="network.log"
  />
  </sharedListeners>
  <trace autoflush="true"/>
</system.diagnostics>
```

The rabbit server log only has:
```
=INFO REPORT==== 20-May-2015::22:00:46 ===
accepting AMQP connection <0.20930.0> (172.16.169.138:2562 -> 172.16.169.1:5672)

=INFO REPORT==== 20-May-2015::22:01:04 ===
closing AMQP connection <0.20930.0> (172.16.169.138:2562 -> 172.16.169.1:5672)
```

**Comment by michaelklishin** [comment]
@mfloryan ok, then it seems to be mostly noise (socket timeouts happen but heartbeat implementation seemingly works correctly nonetheless). I'll look into it.

**Comment by michaelklishin** [comment]
@mfloryan to clarify, is your concern in the fact that exceptions happen or is there actual connectivity loss from the client (or anything else which you find functionally problematic)?

**Comment by mfloryan** [comment]
@michaelklishin I have seen no evidence (in Wireshark) of the connection actually being lost. So on that ground I think we're fine.
It is indeed the number of exceptions that are being thrown that is of some concern. We monitor our application and amongst other things we're monitoring the .Net exceptions and this is creating a level of noise that is less than desirable.

**Comment by michaelklishin** [comment]
Understood, thank you, we will look into handling these exceptions to avoid the scary and unnecessary noise.

**Comment by michaelklishin** [comment]
This should be fixed with #82.
